### PR TITLE
Dinner Tables feature

### PR DESCRIPTION
**Pull request by deepshah**
_Here, tables are physical table at a restaurant, not to be confused with database tables._

Add the option to assign a table to a particular sale. 2 default tables are added- Delivery and Take aways.

### CODE REVIEW

**Comment by daN4cat** [comment]
Hi @deepshah many thanks for this contribution, really good.

We need to review it and think about the best way to merge it in.

I hope you made it configurable so people can enable it only for cases like yours.

**Comment by deepshah** [comment]
Hi @daN4cat 

There is no way to enable/disable it. By default 'Delivery' option is selected. While it shouldn't affect those who do not require it, I will add the option to disable it.

**Comment by daN4cat** [comment]
I see.

I noticed that there is a huge list of files changed but with no real change, to contribute and allow a review to be reasonable you should revert that as merging the code will be a huge effort.

**Comment by daN4cat** [comment]
The PR says that you changed 1478 files!

**Comment by deepshah** [comment]
For some reasons I had changed the file permissions :-|
Have reverted that change

File: application/language/de/bootstrap_tables_lang.php [hunk 1]
```diff
@@ -3,8 +3,8 @@
 $lang["tables_loading"] = "Lade, bitte warten...";
 $lang["tables_rows_per_page"] = "{0} Einträge pro Seite";
 $lang["tables_page_from_to"] = "Zeige {0} bis {1} von {2} Zeile(n)";
-$lang["tables_hide_show_pagination"] = "Seitenzahlen anzeigen/verbergen";
-$lang["tables_refresh"] = "Aktualisieren";
+$lang["tables_hide_show_pagination"] = "Hide/Show pagination";
```
Comment (daN4cat) [inline application/language/de/bootstrap_tables_lang.php]: Please revert these changes to de (German) language, they are not correct

File: application/language/es/common_lang.php [hunk 2]
```diff
@@ -22,7 +22,7 @@
 $lang["common_gender"] = "Género";
 $lang["common_gender_female"] = "F";
 $lang["common_gender_male"] = "M";
-$lang["common_inv"] = "Inv";
+$lang["common_inv"] = "inv";
```
Comment (daN4cat) [inline application/language/es/common_lang.php]: Please revert these changes to es (Spanish) language, they are not correct

**Comment by deepshah** [comment]
Have made the changes

**Comment by daN4cat** [comment]
Looks better but there are still three files under vendor/ that have permission changed

**Comment by deepshah** [comment]
Don't know how that change happened, will fix it. But shouldn't files have permission 644?
Also should we add vendors directory to .gitignore?

**Comment by daN4cat** [comment]
I would leave the permission change to another submission focus on that only.
The vendors for cannot be in .gitignore it's needed and part of the installation.

**Comment by deepshah** [comment]
Okay, have changed it.

**Comment by miprince** [comment]
How can I change the Discount ( percentage to flat amount)..please help me

**Comment by deepshah** [comment]
@daN4cat any update on this?

**Comment by daN4cat** [comment]
Little time I'm afraid and I'm trying to fix a bug to close 3.0.1

File: application/models/Dinner_table.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
+<?php
+class Dinner_table extends CI_Model
```
Comment (daN4cat) [inline application/models/Dinner_table.php]: this class has code with indentation issues

File: application/models/Dinner_table.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
+<?php
```
Comment (daN4cat) [inline application/models/Dinner_table.php]: please check the other models, there should be a generic first line to protect the code

File: application/models/Dinner_table.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        $this->db->from('dinner_tables');  
+        $this->db->where('dinner_table_id', $dinner_table_id);
+        
+        return ($this->db->get()->num_rows() >= 1);
+    }
+    
+
+    public function save(&$table_data, $dinner_table_id) 
+    {
+		$name = $$table_data['name'];
```
Comment (daN4cat) [inline application/models/Dinner_table.php]: double $$ ?

File: application/views/configs/table_config.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
+<?php echo form_open('config/save_tables/', array('id' => 'table_config_form', 'class' => 'form-horizontal')); ?>
```
Comment (daN4cat) [inline application/views/configs/table_config.php]: I was under the impression that this feature could be enabled/disabled.

Or is it left to the permissions?

File: .gitignore [hunk 7]
```diff
@@ -22,3 +22,4 @@ git-svn-diff.py
 *~
 *.~
 *.log
+application/logs/*
```
Comment (daN4cat) [inline .gitignore]: good one, thanks

File: database/database.sql [hunk 8]
```diff
@@ -505,10 +505,12 @@ CREATE TABLE `ospos_sales` (
   `comment` text NOT NULL,
```
Comment (daN4cat) [inline database/database.sql]: You should add to tables.sql first as database and other two files are automatic generated out of that tables.sql file.

**Comment by jlctmaster** [comment]
Hello @deepshah, I am interested in this improvement to the system, and I saw that they already released version 3.1.0 without it, I could support you to correct what you require in case it is needed.

Greetings.

**Comment by daN4cat** [comment]
@jlctmaster 3.1.0 is not release yet. It's under testing and not everything has been merged yet.

With regards to this we should get @deepshah to rebase to the latest and then we should make it optional.
Not everybody uses OSPOS for a restaurant with tables.

**Comment by jlctmaster** [comment]
Ok, I agree, for that reason, should be something configurable as is the billing, you have the option to enable it or not, I am interested in this support because I want to implement the system in a restaurant and having this support, would be a good Beginning, and I could devote time to the mobile version that I am about to develop.

**Comment by daN4cat** [comment]
If you want to take what done, rebase it and making it configurable that would be great. I'm not sure the original author has time to do more as it's been there for some time...

**Comment by daN4cat** [comment]
What's the mobile version? Better responsive UI or a mobile app connected to OSPOS?

**Comment by jlctmaster** [comment]
A mobile app connected to OSPOS

**Comment by jekkos** [comment]
for mobile see also #593, #615

**Comment by daN4cat** [comment]
@jlctmaster do you plan to add a RESTful API to OSPOS? Because I had that in mind but we need to design it properly.
Also we should design with different channels in mind, so the architecture needs to be laid down in the right way.
We cannot use an Enterprise Service Bus to decouple the RESTful API from the logic because it's a small SaaS story, but we can reason around channels like: Mobile and Web application (when we manage to move to a a proper FrontEnd) and maybe Insight, Stock Management.

**Comment by jlctmaster** [comment]
Of course, my intention is to generate a RestFul API to be able to integrate the mobile app with OSPOS, if you want we can work together for this improvement to the system.

**Comment by daN4cat** [comment]
@jlctmaster yes I would prefer to at least work on the design and architecture part of this story. But let's do this after we have closed 3.1.0.
I would need a RESTful API to extract Insight to build a connector from some google sheet stuff for accounting and strategic buying I have setup.
I'm tired of doing manual copy and paste or report exporting and reimporting...

**Comment by jlctmaster** [comment]
Okay, okay, meanwhile I will be working on the configuration of the restaurant that is dynamically to enable it as well as billing.

**Comment by jekkos** [comment]
Would be good to start with a kind of specification in eg. Swagger so that clients can be easily generated

**Comment by jlctmaster** [comment]
Hello, I already have something prepared to upload and to perform the tests, however, I did this in a branch of my fork, should I make a new PR?

**Comment by odiea** [comment]
It appears the he has not added any English language support for his dinner table config.

**Comment by jlctmaster** [comment]
I still have not uploaded the changes I made, I'm asking if I make a new PR or how I would integrate with this PR

**Comment by odiea** [comment]
I downloaded your branch from your git site. If that is not up to date I apologize.

**Comment by jlctmaster** [comment]
Okay, if I need to upload that last change, I'll let you know what I've done.

**Comment by jlctmaster** [comment]
Hello @odiea , already post the changes. Please check it

**Comment by daN4cat** [comment]
@jlctmaster you need to pull the code on your side and go for a new PR from your repo. I will close this one once your one is submitted.

**Comment by daN4cat** [comment]
Replaced by #1152
